nemo/collections/nlp/parts/nlp_overrides.py

```diff
@@ -205,51 +209,184 @@ def init_model_parallel(self, global_rank: int, world_size: int) -> None:
+                model_key = 'model'
+                model_attr = self.lightning_module.model
+            elif hasattr(self.lightning_module, 'enc_dec_model'):
+                model_key = 'enc_dec_model'
+                model_attr = self.lightning_module.enc_dec_model
+            if model_key is not None:
+                if isinstance(model_attr, Float16Module):
+                    new_state_dict = {}
+                    for key in checkpoint['state_dict'].keys():
+                        new_key = key.replace(f'{model_key}.', f'{model_key}.module.', 1)
+                        new_state_dict[new_key] = checkpoint['state_dict'][key]
+                    checkpoint['state_dict'] = new_state_dict
+
+            self.lightning_module.load_state_dict(checkpoint["state_dict"])
+
+    def _fix_tensors_device(self, ckpt: Dict) -> Dict:
+        """ Ensure checkpoint tensors are on the correct device."""
+        assert torch.cuda.is_initialized(), (torch.cuda.is_available(), torch.cuda.is_initialized())
+        cur_dev = torch.device("cuda", index=torch.cuda.current_device())
+
+        def _fix_device(t):
+            if isinstance(t, torch.Tensor) and t.is_cuda and t.device != cur_dev:
+                t = t.to(cur_dev)
+            return t
+
+        return dict_list_map_outplace(_fix_device, ckpt)
 
     def load_checkpoint(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
-        """ PTL override to accomodate model parallel checkpoints """
-        # TODO: move to CheckpointIO
-        torch.cuda.empty_cache()
-        checkpoint_path = inject_model_parallel_rank(checkpoint_path)
-        return self.checkpoint_io.load_checkpoint(checkpoint_path)
+        """ PTL method which we override to integrate distributed checkpoints for model parallel models.
+            In order to load distributed checkpoints we need to provide the sharded_state_dict to 
+            the distributed load function. We get the sharded_state_dict from self.lightning_module
+            which makes it convenient to have the loading logic happen at the strategy level.
+        """
+
+        # Try to read the checkpoint at `path`. If not exist, do not restore checkpoint.
+        fs = get_filesystem(checkpoint_path)
+        if not fs.exists(checkpoint_path):
+            raise FileNotFoundError(f"Checkpoint at {checkpoint_path} not found. Aborting training.")
+
+        # Check if using distributed checkpointing
+        if getattr(self.lightning_module, 'mcore_gpt', False):
+
+            # Distributed checkpoints must be directories.
+            if not fs.isdir(checkpoint_path):
+                raise ValueError(f'Distributed checkpoints should be a directory. Found: {checkpoint_path}.')
+
+            sharded_state_dict = self.lightning_module.sharded_state_dict()
+
+            checkpoint = {}
+
+            # after dist_checkpointing.load, sharded tensors will be replaced with tensors
+            checkpoint['state_dict'] = sharded_state_dict
```

Comment:
I believe we could avoid creating this manually and instead call sth like `checkpoint = self.trainer.checkpoint_connector.dump_checkpoint()`, provided that we override `NLPDDPStrategy.lightning_module_state_dict` to return sharded_state_dict (maybe conditionally) and `NLPDDPStrategy.optimizer_state` to return optimizer_sharded_state_dict.

This will also simplify saving logic a bit

Reply:
Let's add that in a follow up PR.